ddht/rpc.py

```diff
@@ -0,0 +1,253 @@
+
+    if "method" not in request:
+        raise ValidationError("Missing 'method' key")
+    if "params" in request:
+        if not isinstance(request["params"], list):
+            raise ValidationError("Missing 'method' key")
+
+
+def generate_response(
+    request: RPCRequest, result: Any, error: Optional[str]
+) -> RPCResponse:
+    response = RPCResponse(
+        id=request.get("id", -1), jsonrpc=request.get("jsonrpc", "2.0"),
+    )
+
+    if result is None and error is None:
+        raise TypeError("Must supply either result or error for JSON-RPC response")
+    if result is not None and error is not None:
+        raise TypeError(
+            "Must not supply both a result and an error for JSON-RPC response"
+        )
+    elif result is not None:
+        response["result"] = result
+    elif error is not None:
+        response["error"] = str(error)
+    else:
+        raise Exception("Unreachable code path")
+
+    return response
+
+
+TParams = TypeVar("TParams")
+TResult = TypeVar("TResult")
+
+
+class RPCError(Exception):
+    ...
+
+
+class RPCHandler(RPCHandlerAPI, Generic[TParams, TResult]):
+    async def __call__(self, request: RPCRequest) -> RPCResponse:
+        try:
+            params = self.extract_params(request)
+            result = await self.do_call(params)
+        except RPCError as err:
+            return self.generate_error_response(request, str(err))
+        else:
+            return self.generate_success_response(request, result)
+
+    @abstractmethod
+    async def do_call(self, params: TParams) -> TResult:
+        ...
+
+    def extract_params(self, request: RPCRequest) -> TParams:
+        return request.get("params", [])  # type: ignore
+
+    def generate_success_response(
+        self, request: RPCRequest, result: Any
+    ) -> RPCResponse:
+        return generate_response(request, result, None)
+
+    def generate_error_response(self, request: RPCRequest, error: str) -> RPCResponse:
+        return generate_response(request, None, error)
+
+
+class UnknownMethodHandler(RPCHandlerAPI):
+    async def __call__(self, request: RPCRequest) -> RPCResponse:
+        return generate_response(
+            request, None, f"Unknown RPC method: {request['method']}"
+        )
+
+
+fallback_handler = UnknownMethodHandler()
+
+
+class RPCServer(Service):
+    logger = logging.getLogger("alexandria.rpc.RPCServer")
+    _handlers: Dict[str, RPCHandlerAPI]
+
+    def __init__(
+        self, ipc_path: pathlib.Path, handlers: Dict[str, RPCHandlerAPI]
+    ) -> None:
+        self.ipc_path = ipc_path
+        self._handlers = handlers
+        self._serving = trio.Event()
+
+    async def wait_serving(self) -> None:
+        await self._serving.wait()
+
+    async def run(self) -> None:
+        self.manager.run_daemon_task(self.serve, self.ipc_path)
+        try:
+            await self.manager.wait_finished()
+        finally:
+            self.ipc_path.unlink()
```

Comment:
I'm not sure, but maybe this should use `missing_ok = True`? If `self.serve` fails for some reason, maybe it couldn't bind to the given path, then we'll get an exception, and we'll also get a second exception here which might mask the first.